lib/phoenix/endpoint.ex

```diff
@@ -456,7 +456,8 @@ defmodule Phoenix.Endpoint do
 
   defp plug() do
     quote location: :keep do
-      use Plug.Builder, init_mode: Phoenix.plug_init_mode()
+      @before_compile {Phoenix.Endpoint, :__add_fallback_route__}
```

Comment:
I'm not nuts about how I'm (ab)using Plug.Router throughout Endpoint. The main problem is effectively adding a fallback route that is both performant and doesn't use any private parts of Plug.Router's implementation. 

I'm thinking that a `continue_on_no_match` option to Plug.Router may be a good option, which would add a no-op `do_match` clause at the top of the `Plug.Router.__before_compile` macro. This would enable the general purpose use of Plug.Router matches in a non-comprehensive manner, allowing Plug.Router's plugs to be more effectively interleaved with other Plugs. Regardless, this isn't a blocking concern for the other PRs in https://github.com/phoenixframework/phoenix/issues/5003